### PR TITLE
Add session ID and UI request ID headers

### PR DESCRIPTION
**Pull request by idanovo**

### CODE REVIEW

**Review by eladlachmi — CHANGES_REQUESTED** [review]
Overall, it looks good ✌🏻 Got a few minor comments.

File: pkg/gateway/handler.go [hunk 1]
```diff
@@ -106,6 +106,7 @@ func NewHandler(region string, catalog *catalog.Catalog, multipartTracker multip
 	}
 	loggingMiddleware := httputil.LoggingMiddleware(
 		"X-Amz-Request-Id",
+		"X-Amz-Session-Id",
```
[the remark this answers is not in the document]
Comment (idanovo) [reply pkg/gateway/handler.go:109]: So changing it to `Session-Id` is OK

File: pkg/api/serve.go [hunk 2]
```diff
@@ -29,6 +29,7 @@ import (
 
 const (
 	RequestIDHeaderName = "X-Request-ID"
+	SessionIDHeaderName = "X-Session-ID"
```
Comment (arielshaqed) [inline pkg/api/serve.go:32]: Please document how this is different from Request-ID.  I _think_ that they do roughly the same thing.

**Review by arielshaqed — COMMENTED** [review]
Really puzzled by this one, issue did not help.

File: pkg/httputil/logging.go [hunk 3]
```diff
@@ -54,6 +56,30 @@ func RequestID(r *http.Request) (*http.Request, string) {
...
+	UIReqID := ""
+	if resp != nil {
+		UIReqID = resp.(string)
+	}
+	return UIReqID
+}
+
+func SessionID(r *http.Request) (*http.Request, string) {
+	ctx := r.Context()
+	resp := ctx.Value(SessionIDContextKey)
```
Comment (arielshaqed) [inline pkg/httputil/logging.go:71]: `resp` makes me think of "response", I am not sure that name is relevant here.  Maybe
```suggestion
	sessionID := ctx.Value(SessionIDContextKey)
```
?

File: pkg/httputil/logging.go [hunk 4]
```diff
@@ -54,6 +56,30 @@ func RequestID(r *http.Request) (*http.Request, string) {
...
+	var sessionID string
+	if resp == nil {
+		// assign a session ID for this request
+		sessionID = uuid.New().String()
+		r = r.WithContext(context.WithValue(ctx, SessionIDContextKey, sessionID))
```
Comment (arielshaqed) [inline pkg/httputil/logging.go:76]: Why do we need to do this?  IIUC you explained that users can share a Session-ID header between multiple requests to be able to tie their processing together.  But here there _is_ no Session-ID header.  So this session-ID can only identify a single request; how is it different from the Request-ID?  (I would prefer not to have 2 long UUID strings on each request!)

**Review by arielshaqed — COMMENTED** [review]
Obviously cannot object to a user request.  However from your explanation on the issue it is unclear how a _random_ Session-ID would be useful on a request that already has a Request-ID.  Perhaps we can remove this identifier _when the user has not asked for it_?

File: pkg/httputil/logging.go (hunk 4 above)
Comment (idanovo) [reply pkg/httputil/logging.go:76]: mm I kind of agree. It's a little bit questionable who should be responsible for generating this session ID in the first place, the client or the server. @eladlachmi thoughts?

**Pull request closed without merging**